### PR TITLE
avoid Kubelet request API Server anonymously before client certificate bootstrap finished

### PR DESCRIPTION
**Pull request by answer1991**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Kubelet client certificate manager bootstrap certificate in background. Before the client certificate bootstrap finished, Kubelet always try to connect to API Server anonymously.

Thanks to all most Kubelet components writtern as controller mode,  Kubelet will re-try immediately if API Server reject the Kubelet anonymous request. And in most situation, Kubelet will connect to API Server successfully if certificate bootstrap finished and work normally.

However, API Server receives abnormal requests and may misleading cluster admin.

**Which issue(s) this PR fixes**:
None

**Special notes for your reviewer**:

None

**Does this PR introduce a user-facing change?**:

```release-note
avoid Kubelet request API Server anonymously before client certificate bootstrap finished
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by alejandrox1 — COMMENTED** [review]
This looks good!
/ok-to-test
/priority important-longterm

**Comment by answer1991** [comment]
/retest

I think the fix is a little tricky. We may consider add a new function in certificate manager like `CertificateBootstrapDone()` to tell us if certificate bootstrap done.

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -870,6 +870,18 @@ func buildKubeletClientConfig(s *options.KubeletServer, nodeName types.NodeName)
 		klog.V(2).Info("Starting client certificate rotation.")
 		clientCertificateManager.Start()
 
+		// we should wait certificate manager to finish client certificate bootstrap, otherwise Kubelet always try to connect
+		// to API Server anonymously before client certificate bootstrap finished.
+		if err := wait.PollImmediate(100*time.Millisecond, 5*time.Minute, func() (done bool, err error) {
```
Comment (answer1991) [inline cmd/kubelet/app/server.go:875]: After #92786 merged, I would like use context to poll certificate ready

**Comment by answer1991** [comment]
/hold 

After #92786 merged, I will change it to use context to poll certificate ready.

**Comment by answer1991** [comment]
/hold cancel
/assign @liggitt 

Could you help review this? thanks

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -874,6 +874,18 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
 		klog.V(2).Info("Starting client certificate rotation.")
 		clientCertificateManager.Start()
 
+		// we should wait certificate manager to finish client certificate bootstrap, otherwise Kubelet always try to connect
+		// to API Server anonymously before client certificate bootstrap finished.
+		if err := wait.PollImmediateUntil(100*time.Millisecond, func() (done bool, err error) {
```
Comment (liggitt) [inline cmd/kubelet/app/server.go:879]: I'm pretty sure blocking here would prevent a self-hosted kubelet (where the kube-apiserver was run as a static pod by the kubelet) from ever initializing. The function explicitly logs "Client rotation is on, will bootstrap in background"

Comment (liggitt) [reply cmd/kubelet/app/server.go:879]: see https://github.com/kubernetes/kubernetes/pull/71174/files#diff-e0160f895346669c2e4495acfd858ea7R759

**Comment by answer1991** [comment]
@liggitt Thanks for taking a look.

I found if `RotateCertificates` is disable(`--rotate-certificates==false`), then Kubelet will block in initializing before certificate bootstrap finished. see:

https://github.com/kubernetes/kubernetes/blob/70e09f2c24f170f985e4a6d7cf7101e32c97ea2a/cmd/kubelet/app/server.go#L880-L885

**Comment by liggitt** [comment]
> I found if `RotateCertificates` is disable(`--rotate-certificates==false`), then Kubelet will block in initializing before certificate bootstrap finished.

That's a good catch, but we need a fix that moves that scenario (`--rotate-certificates==false --bootstrap-kubeconfig=...`) into a background goroutine, not one that fails and returns an error that would prevent kubelet startup. We clearly also need a test case around that scenario.

**Pull request closed without merging**